src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+
+        /// <inheritdoc cref="ConfigureSchema(IGraphQLBuilder, Action{ISchema})"/>
+        public static IGraphQLBuilder ConfigureSchema(this IGraphQLBuilder builder, Action<ISchema, IServiceProvider> action)
+            => builder.Register<IConfigureSchema>(new ConfigureSchema(action ?? throw new ArgumentNullException(nameof(action))));
+
+        /// <summary>
+        /// Configures an action to run immediately prior to document execution.
+        /// Assumes that the document executer is <see cref="DocumentExecuter"/>, or that it derives from <see cref="DocumentExecuter"/> and calls
+        /// <see cref="DocumentExecuter(IDocumentBuilder, IDocumentValidator, IComplexityAnalyzer, IDocumentCache, System.Collections.Generic.IEnumerable{IConfigureExecution})"/>
+        /// within the constructor.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="ExecutionOptions.RequestServices"/> can be used within the delegate to access the service provider for this execution.
+        /// </remarks>
+        public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Action<ExecutionOptions> action)
+            => builder.Register<IConfigureExecution>(new ConfigureExecution(action ?? throw new ArgumentNullException(nameof(action))));
+        #endregion
+
+        #region - AddValidationRule -
+        /// <summary>
+        /// Registers <typeparamref name="TValidationRule"/> as a singleton within the dependency injection framework
+        /// as <typeparamref name="TValidationRule"/> and as <see cref="IValidationRule"/>.
+        /// Configures document execution to add the validation rule within <see cref="ExecutionOptions.ValidationRules"/>.
+        /// When <paramref name="useForCachedDocuments"/> is <see langword="true"/>, also configures document execution to
+        /// add the validation rule within <see cref="ExecutionOptions.CachedDocumentValidationRules"/>.
+        /// </summary>
+        /// <remarks>
+        /// If <paramref name="useForCachedDocuments"/> is <see langword="true"/>, do not separately install the validation rule within
+        /// your execution code or the validation rule may be run twice for each execution.
+        /// </remarks>
+        public static IGraphQLBuilder AddValidationRule<TValidationRule>(this IGraphQLBuilder builder, bool useForCachedDocuments = false)
+            where TValidationRule : class, IValidationRule
+        {
+            builder.Register<TValidationRule>(ServiceLifetime.Singleton);
+            builder.Register<IValidationRule, TValidationRule>(ServiceLifetime.Singleton);
+            builder.ConfigureExecution(options =>
+            {
+                var rule = options.RequestServices.GetRequiredService<TValidationRule>();
+                options.ValidationRules = (options.ValidationRules ?? DocumentValidator.CoreRules).Append(rule);
```

Comment:
The same problem here with `overconfiguration` in case of shared options object.

Reply:
We could simply, for v4 at least, clone the options class before executing the configuration delegates.

Reply:
I would not bother with that work for v4. V5 issue.

Reply:
#2796